### PR TITLE
add tst_decode.f #101

### LINKED ISSUE #101: add test like the gf2c tst_decode.c
As discussed in email, we have a new test in g2c, tst_decode.c. A fortran translation of this test should be added to g2.

### PR DESCRIPTION
**Pull request by Hang-Lei-NOAA**
This is rewritten from g2c test: tst_decode.c to fortran used in g2 test.

Part of #26.
Fixes #101

### CODE REVIEW

File: tests/tst_decode.f [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+     &  0x40, 0x00, 0x00, 0x00, 0x22, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+     &  0x02, 0x00, 0x60, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x06, 0x01,
+     &  0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+     &  0x00, 0x00, 0x15, 0x05, 0x00, 0x00, 0x00, 0x79, 0x00, 0x00, 0x00, 0x00,
+     &  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x06,
+     &  0x06, 0xff, 0x00, 0x00, 0x00, 0x15, 0x07, 0xff, 0xef, 0xf7, 0xe0, 0x00,
+     &  0x00, 0x00, 0x08, 0x00, 0x01, 0x83, 0x38, 0xee, 0x3f, 0xa7, 0x80, 0x37,
+     &  0x37, 0x37, 0x37 /)
+       
+!    /* clang-format on */
```
Comment (kgerheiser) [inline tests/tst_decode.f]: No reason to have these

File: tests/tst_decode.f [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+     &  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+     &  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+     &  0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
+     &  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+     &  0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1,
+     &  0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1,
+     &  0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1,
+     &  1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1 /)
+      
+!    /* clang-format on */
```
Comment (kgerheiser) [inline tests/tst_decode.f]: And here

**Review by kgerheiser — APPROVED** [review]
Just a comment about the clang format comment, but other than that looks good.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -1,4 +1,12 @@
...
+# NCEPLIBS-g2_UTILS project.
+#
 
-add_executable(test_g2 test_g2.f90)
-target_link_libraries(test_g2 g2_d)
-add_test(test_g2 test_g2)
+function(g2_test name)
+  add_executable(${name} ${name}.f)
+  add_dependencies(${name} g2)
+  target_link_libraries(${name} PRIVATE g2)
```
Comment (kgerheiser) [inline tests/CMakeLists.txt]: This should be `g2_<kind>`

**Review by kgerheiser — CHANGES_REQUESTED** [review]

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -1,4 +1,12 @@
+# This is the cmake build file for the tests directory of the
+# NCEPLIBS-g2_UTILS project.
+#
 
-add_executable(test_g2 test_g2.f90)
-target_link_libraries(test_g2 g2_d)
-add_test(test_g2 test_g2)
+function(g2_test name)
+  add_executable(${name} ${name}.f)
+  add_dependencies(${name} g2)
```
Comment (edwardhartnett) [inline tests/CMakeLists.txt]: Should this be g2_d instead of g2?

**Review by edwardhartnett — APPROVED** [review]

File: tests/CMakeLists.txt (hunk 4 above)
Comment (Hang-Lei-NOAA) [reply tests/CMakeLists.txt]: yes

File: tests/CMakeLists.txt (hunk 3 above)
Comment (Hang-Lei-NOAA) [reply tests/CMakeLists.txt]: yes, g2_d

File: tests/tst_decode.f (hunk 2 above)
Comment (Hang-Lei-NOAA) [reply tests/tst_decode.f]: removed

File: tests/tst_decode.f [hunk 5]
```diff
@@ -59,39 +71,39 @@ program tst_decode
...
-       if (gfld%version .ne. 2) return G2_ERROR
+       if (gfld%version .ne. 2) print *, G2_ERROR
 
-       if (gfld%ndpts .ne. 121) return G2_ERROR    !/* 11x11 grid */
+       if (gfld%ndpts .ne. 121) print *, G2_ERROR    !/* 11x11 grid */
 
        do i =0,(gfld%ndpts-1) 
-        if (gfld%fld(i) .ne. fld_ok(i)) return G2_ERROR
-         i=i+1
+        if (gfld%fld(i) .ne. fld_ok(i)) print *, G2_ERROR
```
Comment (edwardhartnett) [inline tests/tst_decode.f]: Seems like you are just printing the error message, and still returning 0 from the program.

When there is an error the fortran code must call STOP 2 (or any number other than 0).

That will cause the program to exit with a non-zero result, and CMake will then know that the test failed.

**Review by edwardhartnett — CHANGES_REQUESTED** [review]

File: tests/tst_decode.f (hunk 5 above)
Comment (Hang-Lei-NOAA) [reply tests/tst_decode.f]: agree

File: tests/tst_decode.f [hunk 6]
```diff
@@ -0,0 +1,121 @@
+!>
+!> This test is for decoding full grib2 message.
+!> Dusan Jovic, July, 2021
```
Comment (edwardhartnett) [inline tests/tst_decode.f]: Add your name and the date, and mention that this was derived from Dusan's program.

File: tests/tst_decode.f [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+     &  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+     &  0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1,
+     &  0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1,
+     &  0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1,
+     &  1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1 /)
+       
+ 
+        call gb_info(fgrib, fgrib_len, listsec0, listsec1,
+     &  numfields, numlocal, maxlocal, iret)
+       if (iret .ne. 0)  print *, G2_ERROR
```
Comment (edwardhartnett) [inline tests/tst_decode.f]: You need to stop here if iret .ne. 0, right?

File: tests/tst_decode.f (hunk 6 above)
Comment (Hang-Lei-NOAA) [reply tests/tst_decode.f]: Will add this docs.

File: tests/tst_decode.f [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+       do i =0,12
+         if (listsec1(i) .ne. listsec1_ok(i)) then
+          print *, G2_ERROR
+          stop 102
+         endif
+       end do
+
+       if (numfields .ne. 1) print *, G2_ERROR
+       if (numlocal .ne. 0)  print *, G2_ERROR
```
Comment (edwardhartnett) [inline tests/tst_decode.f]: In each case you also need to stop if these values are incorrect.

File: tests/tst_decode.f (hunk 7 above)
Comment (Hang-Lei-NOAA) [reply tests/tst_decode.f]: Yes, I see this and need to modify it

File: tests/tst_decode.f (hunk 8 above)
Comment (Hang-Lei-NOAA) [reply tests/tst_decode.f]: yes

File: tests/tst_decode.f [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+       call gf_getfld(fgrib, fgrib_len, 1, 1, 1, gfld, iret)
+       if (iret .ne. 0)  print *, G2_ERROR
```
Comment (edwardhartnett) [inline tests/tst_decode.f]: Stop if iret .ne. 0

Comment (Hang-Lei-NOAA) [reply tests/tst_decode.f]: yes

File: tests/tst_decode.f [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+       if (gfld%version .ne. 2) then
+         print *, G2_ERROR
+         stop 103
+       endif
+
+       if (gfld%ndpts .ne. 121) then
```
Comment (edwardhartnett) [inline tests/tst_decode.f]: I would just drop the all the print statements and just make each of these one line of code:

if (gfld%ndpts .ne. 121) stop 121

Comment (Hang-Lei-NOAA) [reply tests/tst_decode.f]: fix this

File: tests/tst_decode.f [hunk 11]
```diff
@@ -1,6 +1,7 @@
-!>
-!> This test is for decoding full grib2 message.
-!> Dusan Jovic, July, 2021
+!> @file
```
Comment (edwardhartnett) [inline tests/tst_decode.f]: We don't use doxygen in test code. ;-)

File: tests/tst_decode.f [hunk 12]
```diff
@@ -75,42 +76,43 @@ program tst_decode
...
-       if (iret .ne. 0)  print *, G2_ERROR
+       if (iret .ne. 0) then
+         print *, G2_ERROR
+          stop 101
+       endif
 
        do i =0,2
          if (listsec0(i) .ne. listsec0_ok(i)) then
           print *, G2_ERROR
-          stop 101
```
Comment (edwardhartnett) [inline tests/tst_decode.f]: Why take out these stop calls? They are also needed, right?

File: tests/tst_decode.f (hunk 10 above)
Comment (edwardhartnett) [reply tests/tst_decode.f]: You took out all the stop calls. They are needed. ;-) The print calls are not needed.

File: tests/tst_decode.f (hunk 12 above)
Comment (Hang-Lei-NOAA) [reply tests/tst_decode.f]: They are there, just reorganized.

**Comment by edwardhartnett** [comment]
Now that --output-on-failure has been added to ctest, we see what is failing in the CI:
```
1/1 Test #1: tst_decode .......................***Failed    0.04 sec
 Testing decoding full grib2 message.\n
 gb_info:  Beginning characters GRIB not found.
STOP 101
```

**Comment by edwardhartnett** [comment]
Any progress here? Let's get this test working and in play...

**Pull request closed without merging**